---
title: "general_purpose_depth_viewer"
output: html_document
date: "2024-10-10"
---



## 2.2 Loading Toy Examples

Load required libraries.

```{r}
library(dplyr)
library(glue)
library(stringr)
library(RPlotRNAseq)
library(SCISSOR)
library(dplyr)
library(data.table)
library(ggplot2)
library(Rsamtools)
# library(GenomicAlignments)
library(parallel)
library(GenomicFeatures)
library(GenomicRanges)
library(patchwork)
theme_set(theme_bw())
```

```{r}
library(devtools)
# document()
load_all()
```



```{r}

# get token
# check if token is outdated

# read manifest
# get target region

# curl download retry until successful
# download and process each depth file. find some way to log the finished sample
# stop processing if any of them throws an error
# 
# save final depth file to file as rds 
#
#

```



```{r}
bam_dir = "~/jlee307/HPV_SV/ArtistR_project"

bamfile_list = 
  list.files(bam_dir,recursive = TRUE,full.names = TRUE) %>%
  grep(pattern="bam$",value = TRUE)

sample_sheet_fn = list.files(bam_dir,pattern = "gdc_sample_sheet",full.names = TRUE)
sample_sheet = read.csv(sample_sheet_fn,sep="\t")


target_coord = "chr5:51,389,204-51,390,104"
target_coord2 = "chr5:51,387,338-51,387,864"


N_cores = 10L

system.time({
mtrx_samtools = 
  get_depth_matrix_gp(
    bam_files = bamfile_list,coord = target_coord
    ,mode = "samtools_custom"
    ,N_cores = N_cores
    ,min_mapq = 30
    ,tmpdir="tmpdir"
    ,modules="samtools/samtools_bcftools_1.20__bedtools_2.31.1"
  )
})


system.time({
mtrx_samtools2 = 
  get_depth_matrix_gp(
    bam_files = bamfile_list,coord = target_coord2
    ,mode = "samtools_custom"
    ,N_cores = N_cores
    ,min_mapq = 30
    ,tmpdir="tmpdir"
    ,modules="samtools/samtools_bcftools_1.20__bedtools_2.31.1"
  )
})

sample_types = 
dplyr::left_join(
  data.frame(File.Name=colnames(mtrx_samtools))
  ,sample_sheet
) %>%
  with(Sample.Type)


```


```{r}
library(ComplexHeatmap)

Heatmap(mtrx_samtools,cluster_rows = FALSE)


depth_hist(mtrx_samtools,smry_fun = max,th=10)
depth_hist(mtrx_samtools2,smry_fun = max,th=10)


mtrx_samtools_filt = filt_samples(mtrx_samtools,th=10)
mtrx_samtools2_filt = filt_samples(mtrx_samtools2,th=10)

mx = rbind(mtrx_samtools,mtrx_samtools2) %>% apply(2,max)

# hist(mx %>% log10)
mtrx_samtools_filt = mtrx_samtools #[,mx>=10]
mtrx_samtools2_filt = mtrx_samtools2 #[,mx>=10]

sample_types_filt = sample_types #[mx>=10]
```


```{r}
exon_coord = "chr5:51,389,646-51,389,932"
exon_coord2 = "chr5:51,387,490-51,387,749"

target_coord_lst = coord_to_lst(target_coord)
target_coord2_lst = coord_to_lst(target_coord2)
exon_coord_lst = coord_to_lst(exon_coord)
exon_coord2_lst = coord_to_lst(exon_coord2)
target_coord_grng = coord_to_grng(target_coord)
target_coord2_grng = coord_to_grng(target_coord2)

exon_coord_lst__intarget =
  exon_coord_lst %>%
  within({
    chr = "target"
    start = start - target_coord_lst$start
    end = end - target_coord_lst$start
  })

exon_coord2_lst__intarget =
  exon_coord2_lst %>%
  within({
    chr = "target"
    start = start - target_coord2_lst$start
    end = end - target_coord2_lst$start
  })


# HeatmapAnnotation(width = )
tant_exon = 
  HeatmapAnnotation(
    Exon = seq_len(width(target_coord_grng)) %in% with(exon_coord_lst__intarget,start:end),
    col = list(Exon = c("TRUE"="black","FALSE"="ivory")),
    height=unit(1,"in")
    
  )
tant_exon2 = 
  HeatmapAnnotation(
    Exon = seq_len(width(target_coord2_grng)) %in% with(exon_coord2_lst__intarget,start:end),
    col = list(Exon = c("TRUE"="black","FALSE"="ivory")),
    height=unit(1,"in")
    
  )

tant_exon_mrg = 
  HeatmapAnnotation(
    Exon = 
      c(
        seq_len(width(target_coord_grng)) %in% with(exon_coord_lst__intarget,start:end),
        seq_len(width(target_coord2_grng)) %in% with(exon_coord2_lst__intarget,start:end)
      )
      ,
    col = list(Exon = c("TRUE"="black","FALSE"="ivory")),
    height=unit(1,"in"),
    border=TRUE
  )


# raw_mtrx = 
#   mtrx_samtools_filt %>%
#   t
# norm_mtrx = 
#   mtrx_samtools_filt %>%
#     apply(2,\(x) x/sum(x) ) %>%
#     apply(2,\(x) log10(x+1)) %>%
#     t
# 
# 
# norm_Q3_mtrx = 
#   mtrx_samtools_filt %>%
#     apply(2,\(x) x/quantile(x,0.75) ) %>%
#     apply(2,\(x) log10(x+1)) %>%
#     t
# 
# 
# norm_max_mtrx = 
#   mtrx_samtools_filt %>%
#     apply(2,\(x) x/max(x) ) %>%
#     # apply(2,\(x) log10(x+1)) %>%
#     t


mx = rbind(mtrx_samtools_filt,mtrx_samtools2_filt) %>% apply(2,max) %>% pmax(10)

norm_max_mtrx = 
  mtrx_samtools_filt %>%
    # apply(2,\(x) x/max(x) ) %>%
    # apply(2,\(x) log10(x+1)) %>%
    t %>%
  {./mx} #%>%
  # apply(1,\(x) log10(x+min(.[.!=0]))) %>%
  # t

norm_max_mtrx2 = 
  mtrx_samtools2_filt %>%
    # apply(2,\(x) x/max(x) ) %>%
    # apply(2,\(x) log10(x+1)) %>%
    t %>%
  {./mx} #%>%
  # apply(1,\(x) log10(x+min(.[.!=0]))) %>%
  # t

# 
# ht = Heatmap(
#   norm_max_mtrx %>% pmax(0.03),
#   cluster_columns = FALSE,
#   show_row_names = FALSE,
#   top_annotation = tant_exon,
#   col = viridis_pal(option="turbo")(11),
#   row_order = order(norm_max_mtrx[,435])
#   )
# 
# ht
# 
# ht = Heatmap(
#   norm_max_mtrx %>% pmax(0.03),
#   cluster_columns = FALSE,
#   show_row_names = FALSE,
#   top_annotation = tant_exon,
#   col = viridis_pal(option="turbo")(11),
#   row_order = order(norm_max_mtrx[,435])
#   )
# 
# ht
# 
# 
# cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0] %>% log10  %>% hist(breaks=30)
# cbind(norm_max_mtrx,norm_max_mtrx2) %>% hist(breaks=30)
```


```{r}
clip_th = 0.02
LQ_prop = 0.25
UQ_prop = 0.9
Gmin = min(cbind(norm_max_mtrx,norm_max_mtrx2))
LQ_S = cbind(norm_max_mtrx,norm_max_mtrx2) %>%
  {quantile(.[.>clip_th],LQ_prop)}
  # apply(2,\(x) quantile(x[x>clip_th],LQ_prop) )  %>% median(na.rm=T)
  # apply(2,\(x) quantile(x[x>min(.)],0.4) )  %>% quantile(0.4,na.rm=T)
median = cbind(norm_max_mtrx,norm_max_mtrx2) %>%
  {median(.[.>clip_th])}
  # apply(2,\(x) quantile(x[x>0.03],0.5) )  %>% quantile(0.5,na.rm=T)
  # apply(2,\(x) quantile(x[x>min(.)],0.5) )  %>% quantile(0.5,na.rm=T)
UQ_S = cbind(norm_max_mtrx,norm_max_mtrx2) %>%
  {quantile(.[.>clip_th],UQ_prop)}
  # apply(2,\(x) quantile(x[x>min(.)],0.9) )  %>% quantile(0.9,na.rm=T)
Gmax = max(cbind(norm_max_mtrx,norm_max_mtrx2))
# min_Smax = X_scaled_log2p1 %>% apply(2,\(x) c(min(x),max(x))) %>% .[2,] %>% min # min of sample max



cntr = median
col_x_scaled =
  scales::viridis_pal(option="turbo")(11) %>%
  {circlize::colorRamp2(
    c(
      Gmin-cntr,
      (LQ_S-cntr)*((length(.)%/%2):1)/(length(.)%/%2),
      0,
      (UQ_S-cntr)*(1:(length(.)%/%2))/(length(.)%/%2),
      Gmax-cntr
    ) +cntr
    ,c(.[1],.,tail(.,n=1)))}

cbind(norm_max_mtrx,norm_max_mtrx2) %>% log10  %>% hist(breaks=30);



nzmin =cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0] %>% min
nzmax =cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0] %>% max

col_pos_vec = seq(log10(nzmin/1.5),log10(nzmax*1.5),length=100)
# col_pos_vec = c(col_pos_vec[1]-diff(col_pos_vec)[1],col_pos_vec,tail(col_pos_vec,n=1)+diff(col_pos_vec)[1])


col_pos = data.frame(x=-10:110/100)
# col_pos = data.frame(x=10^col_pos_vec)
ggplot() +
  geom_col(
    # data = col_pos %>% dplyr::filter(x!=0),aes(x=x,y=Inf,fill=glue("c{x}") ),
    # width = min(diff(log10(col_pos$x))),
    data = col_pos,aes(x=x,y=Inf,fill=glue("c{x}") ),
    width = min(diff((col_pos$x))),
    show.legend = FALSE) +
  scale_fill_manual(values = 
                      with(col_pos,structure(
                        col_x_scaled(x),
                        names=glue("c{x}")
                      ))
                    ) +
  geom_histogram(
    # data = data.frame(x = as.vector(cbind(norm_max_mtrx,norm_max_mtrx2))) %>% dplyr::filter(x!=0),
    data = data.frame(x = as.vector(cbind(norm_max_mtrx,norm_max_mtrx2))),
                 aes(x=x),col="black",alpha=0.5,fill="white") #+
  # scale_x_log10()




ht = Heatmap(
  cbind(norm_max_mtrx,norm_max_mtrx2)[sample_types_filt!="Solid Tissue Normal",]
  #%>% pmax(quantile(cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0],0.4))
  ,
  column_split =c(rep("Exon4",NCOL(norm_max_mtrx)),rep("Exon3",NCOL(norm_max_mtrx2))),
  cluster_column_slices = FALSE,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  top_annotation = tant_exon_mrg,
  col = col_x_scaled#,
  # row_order = order(norm_max_mtrx[,435])
  )

# ht

ht_N = Heatmap(
  cbind(norm_max_mtrx,norm_max_mtrx2)[sample_types_filt=="Solid Tissue Normal",]
  #%>% pmax(quantile(cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0],0.4))
  ,
  column_split =c(rep("Exon4",NCOL(norm_max_mtrx)),rep("Exon3",NCOL(norm_max_mtrx2))),
  cluster_column_slices = FALSE,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  # top_annotation = tant_exon_mrg,
  col = col_x_scaled,
  height = unit(1,"in")
  # row_order = order(norm_max_mtrx[,435])
  )

ht %v% ht_N
# 
# ht2 = Heatmap(
#   norm_max_mtrx2 
#   #%>% pmax(quantile(cbind(norm_max_mtrx,norm_max_mtrx2) %>% .[.!=0],0.4))
#   ,
#   cluster_columns = FALSE,
#   show_row_names = FALSE,
#   top_annotation = tant_exon2,
#   col = col_x_scaled #,
#   # row_order = order(norm_max_mtrx[,435])
#   )
# 
# (ht2 + ht) %v% (ht2 + ht)

# ht = Heatmap(
#   raw_mtrx,
#   cluster_columns = FALSE,
#   show_row_names = FALSE,
#   top_annotation = tant_exon,
#   col = viridis_pal(option="turbo")(11) #,
#   # row_order = order(raw_mtrx[,435])
#   )
# 
# ht




```


```{r}

which(
  seq_len(width(target_coord_grng)) %in% with(exon_coord_lst__intarget,start:end)
)

```






```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
